### ISSUE TITLE
I want to know the cache function be verified by restarting only one client?

### ISSUE THREAD

**Issue by LuJiaWeiwei** [eclipse-openj9/openj9#17027]
After JITServer is enabled, can the cache function be verified by restarting only one client? If a client is restarted, will its cache on the server be deleted, so the cache function will not work?

**Comment by pshipton** [comment]
@mpirvu

**Comment by mpirvu** [comment]
@LuJiaWeiwei The JITServer has two major types of caches: (1) The JITServer AOT cache which holds previously compiled methods in relocatable format and (2) the session cache (one for each client) that holds meta-data specific to each client.
First, I would like to understand which type of cache you are referring to. 
If you are talking about the JITServer AOT cache, to verify its functionality you need to 
(1) start JITServer (AOT cache will be empty at this point) 
(2) start a client and attach it to the JITServer. Most of the compilations requested by the client will be cached at the server. However, this first client will see no benefit
(3) start a new client. This one can take advantage of the populated AOT cache

To verify that JITServer is serving methods from its cache, enable verbose logs at the client with `-Xjit:verbose={compilePerformance},vlog=myvlogs.txt` . Lines that start with `+` represent successful compilations. If the compilation was sent to the server, that line will include the word "remote". If that compilation was served from the JITServer AOT cache, that line will also include the word "deserialized".

The JITServer AOT cache gets deleted when the JITServer instance gets destroyed, unless AOT cache persistence is enabled.

**Comment by LuJiaWeiwei** [comment]
@mpirvu  Thanks for your reply. I add this option, now I can see "+" and "remote", but there is no  "deserialized".
`+ (cold) com/fasterxml/jackson/databind/deser/impl/MethodProperty.deserializeAndSet(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;Ljava/lang/Object;)V @ 00007F46F48FCDA8-00007F46F48FD4C7 OrdinaryMethod I Q_SZ=169 Q_SZI=169 QW=922 j9m=00000000006A93A0 bcsz=127 sync time=3344us mem=[region=1408 system=2048]KB compThreadID=1 CpuLoad=0%(0%avg) JvmCpu=482% queueTime=4977us
 (cold) Compiling com/fasterxml/jackson/databind/deser/std/StringDeserializer.deserialize(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/String;  OrdinaryMethod j9m=00000000006A07A8  t=148747 compThreadID=2 memLimit=262144 KB freePhysicalMemory=8976 MB
+ (cold) com/fasterxml/jackson/databind/deser/std/StringDeserializer.deserialize(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/String; @ 00007F46F48FD668-00007F46F48FDABF OrdinaryMethod I Q_SZ=168 Q_SZI=168 QW=916 j9m=00000000006A07A8 bcsz=122 sync time=1762us mem=[region=1216 system=2048]KB compThreadID=2 CpuLoad=0%(0%avg) JvmCpu=482% queueTime=2422us
`
I start server with "./jitserver -XX:+JITServerUseAOTCache -Xshareclasses -XX:JITServerAOTmx=5000M",
start client with " -Xshareclasses -XX:+JITServerUseAOTCache -XX:+UseJITServer -XX:JITServerAddress={my ip address} -XX:+RequireJITServer"
can you tell me what I need to add

**Comment by mpirvu** [comment]
There is no need to add `-Xshareclasses` at the server. Only the client needs that option.
I just tried with Java17 release 0.36.0 of OpenJ9 using just "java -version" as a very small test program.

1. Start the server with `$JAVA_HOME/bin/jitserver -XX:+JITServerLogConnections -XX:+JITServerUseAOTCache`
2. Destroy possible existing shared class cache at the client: `$JAVA_HOME/bin/java -Xshareclasses:name=myscc,destroy`
3. Run the client which will populate the AOT cache (and the local shared class cache):  `$JAVA_HOME/bin/java -XX:+UseJITServer -XX:+JITServerUseAOTCache -Xshareclasses:name=myscc -Xjit:verbose={compilePerformance},vlog=vlog.txt,disableAsyncCompilation --version`
4. Destroy the local shared class cache created/populated by the client. Otherwise, the client will load AOT compiled bodies from this shared class cache instead of going to the server, because it's faster that way: `$JAVA_HOME/bin/java -Xshareclasses:name=myscc,destroy`
5. Run the client again: `$JAVA_HOME/bin/java -XX:+UseJITServer -XX:+JITServerUseAOTCache -Xshareclasses:name=myscc -Xjit:verbose={compilePerformance},vlog=vlog.txt,disableAsyncCompilation --version` This time the client will issue remote compilation requests and some of them will be satisfied from the remote AOT cache.
6. Look for "deserialized" in the verbose log created by the second invocation of the client.

Notes:
- Above I used `disableAsyncCompilation` for the clients to make the application threads wait for compilation results before proceeding. This is needed here because "java -version" is such a small program, it may end before compilations are done. For a regularly sized Java program you should NOT use `disableAsyncCompilation`.
- Above I didn't specify the address of the server with `-XX:JITServerAddress=...` because both the client and the server ran on the same machine, so we can use the default `localhost`. You should add that option if the server is remote or the client runs in containers.
- The local shared class cache (SCC) caches AOT methods much like the JITServer AOT cache. The client will use the local SCC in preference of the remote JITServer AOT cache because it is faster. The JITServer AOT cache is useful when running JVMs in containers and the local SCC that the client JVM creates/populates is not shared with other containers (and it gets destroyed when the container terminates).

**Comment by LuJiaWeiwei** [comment]
@mpirvu  Thanks for your reply. But it seems that local shared class cache is not the problem. It can be observed clearly the compilation is offloaded to server. But the server cpu fluctuation is almost same when start another client with same program, in my understand, the server's burden will be less if the AOT cache is used in right way when I start another client.

**Comment by mpirvu** [comment]
>  the server's burden will be less if the AOT cache is used in right way when I start another client.

Correct. However, according to your comment https://github.com/eclipse-openj9/openj9/issues/17027#issuecomment-1486144943, something is off in your setup such that compiled methods are not fetched from the JITServer AOT cache (missing "deserialized" keyword in the verbose log), and I was trying to get to the bottom of this before proceeding further.

Are you running in containers or directly on a machine/VM?

**Comment by LuJiaWeiwei** [comment]
@mpirvu  I run clients on k8s and run server directly on a machine.

**Comment by mpirvu** [comment]
k8s means containers. Where do you create the shared class cache for the clients, in a volume to be shared by all clients, or embedded in each container?

**Comment by LuJiaWeiwei** [comment]
@mpirvu  emmm, I think the cache will be stored in server at first. What should I do to share cache for  clients in different containers.

**Comment by LuJiaWeiwei** [comment]
@mpirvu  If containers are run on different machines, how can I share a directory in different containers. Is it possible.

**Comment by LuJiaWeiwei** [comment]
![image](https://user-images.githubusercontent.com/25796126/228420727-ccc343f4-5931-4579-a090-c150586b264d.png)
I record the CPU fluctuation of the server, the result seem that only AOT cache can benefit little. I have to make the shared class cache work.

**Comment by mpirvu** [comment]
Note that some of the AOT methods will be recompiled to boost throughput. Those recompilations are going to consume CPU at the server.

**Comment by LuJiaWeiwei** [comment]
@mpirvu  Is there some option to disable such recompilation.

**Comment by mpirvu** [comment]
> Is there some option to disable such recompilation.

I am not sure this is desirable, but you can use `-Xjit:inhibitRecompilation` at the client. Note that if you have multiple Xjit options you have to put them under the same `-Xjit:` heading, separated by comma.

If CPU consumption is your main concern, another option to try is `-Xtune:virtualized` (at the client). Both my suggestions will have some effect on peak throughput.
